File: Problems/RateLimiter/RateLimiterFullFeatured.java
// Author: Tushar Jaiswal
// Creation Date: 2025-07-13

// RateLimiter which implements a sliding time window where requests are allowed up to the configured number.

// This supports multiple time windows such as the following.
// Allow:
//     max 1 request per second
//     max 5 requests per minute
//     max 100 requests per hour

//  Support for bursting and cooldown
//  1. Bursting: Allow more requests in a short time than the steady rate (e.g., allow 5 instant requests, then enforce 1/s rate).
//  2. Cooldown: If a client exceeds the limit, block further requests for a cooldown period (e.g., 10 seconds).

//  Runtime Complexity: Each shouldAllowRequest check is O(|Rules| * MaxAllowedRequestsAmongstRules)
//  Space Complexity: Map is O(|Clients| * |Rules| * |MaxAllowedRequestsAmongstRules|)

// If we assume, requests come in chronological order, we can use Deque instead of Heap for faster operations

public class RateLimiter {
    private final List<RateLimitRule> rules;
    
    // client -> window size -> queue of timestamps
    private final Map<String, Map<Integer, Deque<Integer>>> clientWindowMap;
    
    // client -> window size -> cooldown_end_timestamp
    private final Map<String, Map<Integer, Integer>> clientCooldownMap;

    
    public RateLimiter(List<RateLimitRule> rules) {
        this.rules = rules;
        this.clientWindowMap = new HashMap<>();
        this.clientCooldownMap = new HashMap<>();
    }

    public boolean shouldAllowRequest(Request request) {
        String client = request.client;
        int timestamp = request.timestamp;

        clientWindowMap.putIfAbsent(client, new HashMap<>());
        clientCooldownMap.putIfAbsent(client, new HashMap<>());
        Map<Integer, Deque<Integer>> windowMap = clientWindowMap.get(client);
        Map<Integer, Integer> cooldownMap = clientCooldownMap.get(client);

        for (RateLimitRule rule : rules) {
            int window = rule.windowSize;
            int burst = rule.burstCapacity;
            Integer cooldown = rule.cooldownPeriod;

            // Check if in cooldown
            Integer cooldownUntil = cooldownMap.get(window);
            if (cooldownUntil != null && timestamp < cooldownUntil) {
                return false;
            }

            windowMap.putIfAbsent(window, new LinkedList<>());
            Deque<Integer> q = windowMap.get(window);

            // Remove expired entries
            while (!q.isEmpty() && timestamp - q.peekFirst() >= window) {
                q.pollFirst();
            }

            if (q.size() >= burst) {
                // If cooldown is configured, set it
                if (cooldown != null) {
                    cooldownMap.put(window, timestamp + cooldown);
                }
                return false;
            }
        }

        // Passed all rules, add timestamp to queues
        for (RateLimitRule rule : rules) {
            int window = rule.windowSize;
            windowMap.get(window).addLast(timestamp);
        }

        return true;
    }

    // === Example usage ===
    public static void main(String[] args) {
        List<RateLimitRule> rules = Arrays.asList(
            new RateLimitRule(1, 1, 2, 5),      // 2 burst/sec, cooldown 5s
            new RateLimitRule(60, 10, 15, 10)   // 15 burst/min, cooldown 10s
        );

        RateLimiter limiter = new RateLimiter(rules);

        List<Request> requests = Arrays.asList(
            new Request("user1", 1),
            new Request("user1", 1),
            new Request("user1", 1),  // should block + cooldown
            new Request("user1", 2),  // still blocked (cooldown)
            new Request("user1", 7),  // after cooldown
            new Request("user1", 8)
        );

        for (Request r : requests) {
            boolean allowed = limiter.shouldAllowRequest(r);
            System.out.println(String.format("Request at time %d from %s is %s",
                r.timestamp, r.client, allowed ? "allowed" : "blocked"));
        }
    }
}

class Request {
    String client;
    int timestamp;

    public Request(String client, int timestamp) {
        this.client = client;
        this.timestamp = timestamp;
    }
}

class RateLimitRule {
    int windowSize;        // in seconds
    int maxRequests;
    int burstCapacity;
    Integer cooldownPeriod;  // in seconds (nullable)

    public RateLimitRule(int windowSize, int maxRequests, Integer burstCapacity, Integer cooldownPeriod) {
        this.windowSize = windowSize;
        this.maxRequests = maxRequests;
        this.burstCapacity = (burstCapacity != null) ? burstCapacity : maxRequests;
        this.cooldownPeriod = cooldownPeriod;
    }
}

File: Problems/RateLimiter/RateLimiterFullFeatured.py
# Author: Tushar Jaiswal
# Creation Date: 2025-07-13

# RateLimiter which implements a sliding time window where requests are allowed up to the configured number.

# This supports multiple time windows such as the following.
# Allow:
#     max 1 request per second
#     max 5 requests per minute
#     max 100 requests per hour

# Support for bursting and cooldown
# 1. Bursting: Allow more requests in a short time than the steady rate (e.g., allow 5 instant requests, then enforce 1/s rate).
# 2. Cooldown: If a client exceeds the limit, block further requests for a cooldown period (e.g., 10 seconds).

# Runtime Complexity: Each shouldAllowRequest check is O(|Rules| * MaxAllowedRequestsAmongstRules)
# Space Complexity: Map is O(|Clients| * |Rules| * |MaxAllowedRequestsAmongstRules|)

# If we assume, requests come in chronological order, we can use Deque instead of Heap for faster operations

from collections import defaultdict, deque

class Request:
    def __init__(self, client, timestamp):
        self.client = client
        self.timestamp = timestamp
        # Other info like requestID useful in realWorld but not needed for rateLimiter functionality
        
class RateLimitRule:
    def __init__(self, window_size, max_requests, burst_capacity=None, cooldown_period=None):
        self.window_size = window_size                # In seconds
        self.max_requests = max_requests              # Steady rate
        self.burst_capacity = burst_capacity or max_requests
        self.cooldown_period = cooldown_period        # Optional cooldown in seconds
    
class RateLimiter:
    def __init__(self, rules):
        """
        rules: list of RateLimitRule objects
        """
        self.rules = rules
        
        # client -> window -> deque[timestamps]
        self.client_window_map = defaultdict(lambda: defaultdict(deque))  
        
        # client -> window -> cooldown_end_timestamp
        self.client_cooldown_map = defaultdict(dict)  


    def should_allow_request(self, request):
        client = request.client
        timestamp = request.timestamp

        for rule in self.rules:
            window = rule.window_size
            max_requests = rule.max_requests
            burst_capacity = rule.burst_capacity
            cooldown = rule.cooldown_period

            # Check cooldown
            cooldown_until = self.client_cooldown_map[client].get(window)
            if cooldown_until and timestamp < cooldown_until:
                return False  # In cooldown

            q = self.client_window_map[client][window]

            # Remove expired requests
            while q and timestamp - q[0] >= window:
                q.popleft()

            if len(q) >= burst_capacity:
                # Trigger cooldown if configured
                if cooldown:
                    self.client_cooldown_map[client][window] = timestamp + cooldown
                return False

        # Passed all checks; add timestamp
        for rule in self.rules:
            window = rule.window_size
            self.client_window_map[client][window].append(timestamp)

        return True


# Example usage
if __name__ == "__main__":
    rules = [
        RateLimitRule(window_size=1, max_requests=1, burst_capacity=2, cooldown_period=5),     # 2 requests/sec burst, cooldown 5s
        RateLimitRule(window_size=60, max_requests=10, burst_capacity=15, cooldown_period=10), # burst of 15/min, 10s cooldown
    ]

    limiter = RateLimiter(rules)

    test_requests = [
        Request("user1", 1),
        Request("user1", 1),
        Request("user1", 1),
        Request("user1", 2),  # In cooldown
        Request("user1", 7),  # After cooldown
        Request("user1", 8),
    ]

    for r in test_requests:
        result = limiter.should_allow_request(r)
        print(f"Request at time {r.timestamp} from {r.client} is {'allowed' if result else 'blocked'}")
